### ISSUE TITLE
Can we make winapi optional?

### ISSUE THREAD

**Issue by yutannihilation** [extendr/libR-sys#107]
`winapi` is needed because the output of R was not UTF-8 on Windows, and that was a problem (https://github.com/extendr/libR-sys/pull/18#issuecomment-741873561). But, it seems winapi is too heavy to include only for this purpose. I think

* The default encoding of R >= 4.2 is UTF-8 on Windows (but the default of the OS might not, so I'm not sure if it's really safe to assume that `rout` is UTF-8)
* But, at least we can set `LIBRSYS_R_VERSION` so that we don't need to communicate the output of `R`

```
$ du -sh vendor/* | grep M
5.5M    vendor/libR-sys
3.7M    vendor/libc
2.1M    vendor/syn
7.6M    vendor/winapi
53M     vendor/winapi-i686-pc-windows-gnu
55M     vendor/winapi-x86_64-pc-windows-gnu
```

**Comment by Ilia-Kosenkov** [comment]
It seems we use like two functions from `kernel32`. Maybe we can just FFI into it directly?
Also, we can inspect the codepage first -- if it is [65001](https://docs.microsoft.com/en-us/windows/win32/intl/code-page-identifiers) (`utf-8`), we can re-interpret the output as Rust string. Otherwise, we can try to directly call `kernel32`'s [`MultiByteToWideChar`](https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-multibytetowidechar).

Here I re-wrote the same thing. I believe it should work as is on any version of Windows. Separate testing is needed for ARM version if we ever want to run something there (mainly because of the calling convention). 
```rust
use std::os::windows::prelude::OsStringExt;

fn main() {
    let cp = unsafe { get_console_code_page() };
    assert_eq!(cp, 65001); // Testing utf-8 case

    let msg = "Hello, world!"; // This message is `utf-8`, so we can convert it into `utf-16` and `OsString`

    let wide_len = unsafe { multi_byte_to_wide_char(cp, 0, msg.as_ptr(), msg.len() as i32, std::ptr::null_mut(), 0) };
    let mut wide_str = Vec::with_capacity(wide_len as usize);
    let len_written = unsafe { multi_byte_to_wide_char(cp, 0, msg.as_ptr(), msg.len() as i32, wide_str.as_mut_ptr(), wide_len)};
     assert_eq!(len_written, wide_len);
    unsafe {wide_str.set_len(len_written as usize)};
  
    let os_str = std::ffi::OsString::from_wide(&wide_str);

    println!("{os_str:?}");
}

#[link(name = "kernel32")]
extern "system" {
    #[link_name = "GetConsoleCP"]
    fn get_console_code_page() -> u32;
    #[link_name = "MultiByteToWideChar"]
    fn multi_byte_to_wide_char(CodePage: u32, dwFlags : u32, lpMultiByteStr : *const u8, cbMultiByte : i32, lpWideCharStr : *mut u16, cchWideChar : i32) -> i32;
}
```
There is no extra memory management here, or Windows-specific things: just pass a `u8` buffer and a `u16` buffer with given sizes.

**Comment by yutannihilation** [comment]
Sounds good to me!

**Comment by Ilia-Kosenkov** [comment]
Here is my POC branch https://github.com/Ilia-Kosenkov/libR-sys/tree/remove-winapi-dependency and the tests seem to pass https://github.com/Ilia-Kosenkov/libR-sys/actions/runs/2802713112

**Comment by yutannihilation** [comment]
Thanks! Could you take this over when you get time? It's not in a hurry. I'll stick with https://github.com/extendr/libR-sys/pull/108 at least for my recent submission to CRAN anyway because it looks safer at the moment (but I'm definitely not happy with adding another complexity).

**Comment by Ilia-Kosenkov** [comment]
Before I do, can you check your workflow against my branch? Just to double check that the winapi dependency is gone.

**Comment by yutannihilation** [comment]
Confirmed it works as expected.

https://github.com/yutannihilation/string2path/pull/38

**Comment by yutannihilation** [comment]
Fixed by #109
